### ISSUE TITLE
Feature: Certificate to Application Mapping Information in the metrics

### ISSUE THREAD

**Issue by AnuragEkkati** [enix/x509-certificate-exporter#34]
Hi

We explored multiple tools to monitor the certificates on k8s. But one feature which is missing everywhere is the certificate to application mapping. 

Say a given certificate is expired or going to expire, it will be great if we know which application will be affected by it. 

I can create a PR for this if the maintainers feel it is a good feature to have

**Comment by npdgm** [comment]
Hi

Indeed identification of affected workloads may be needed to process alerts, for example when dealing with multi-tenant clusters for alerts to be routed to the correct user or team. Usually this would likely be handled by Alertmanager with routing rules based on namespace. However it may not offer the granularity you're looking for because many applications are deployed in the same namespace.

So I do understand the need and I'm curious how the exporter would be providing the missing link for a solid mapping.
It would be a neet feature if done right!

What implementation do you have in mind?
Scanning all Kubernetes resources to find backreferences to Secrets would be really computational intensive for the exporter and the apiserver. While not being exhaustive enough because of unexpected Secret consumers: custom resources, operators, service meshes, ingress controllers not using Ingress, etc.

Cheers

**Comment by AnuragEkkati** [comment]
Yes, I am doing the same, to scan kubernetes resources ( deployments, statefulsets, daemonsets ) and embed secret_reference information into the metrics. I agree it is computational intensive, but i thought we can control the periodicity of the computation. Or we can have a sidecar which can compute this while the main container serve this information.

I haven't tested the implementation on cluster with many deployments, statefulsets or daemonsets.

**Comment by npdgm** [comment]
Considering this would be an optional feature enabled with a flag, so that nobody gets the additional load on their cluster by default:
I have two concerns.

First is with the restitution of this application mapping on metric labels.
One issue this exporter is already dealing with is avoiding metric duplicates while keeping label cardinality low to behave nicely with Prometheus, preventing storage growth over time with a stable number of certificate. That's best practice, and what default settings are expected to do. It can be tuned anyway, and I remember an other feature was even added with regards to saving storage for a notoriously expensive SaaS platform.
Propagating application mapping to a certificate metric could consequently require new label(s) or metric(s).
The first issue lays with dealing with Secrets referenced by many resources. Finding a way to communicate all back-references. For example a unique label could hold a list of kinds and names, or one label for each kind having a list of names. This is likely what you put in `secret_reference`. Unfortunately it wont create a stable metric cardinality as an unmodified certificate will see label changes while application deployments are altered. Creating new metrics in storage for the same certificate. This is bad practice territory.
A better way could be adding new metric names to add sorts of pointers to used Secrets. A PromQL query is then able to match a certificate with applications. One problem with that is a lack of unique identifier for joining the query. Perhaps this could be solved though.
A second issue is having a solid way to never yield duplicates. I think these two examples with labels or reference metrics are safe, other implementations are not.

My second concern is about the computation issue, and the effectiveness of the feature. Keeping in mind the ease and cost of maintaining this piece of software while keeping the broadest compatibility across all versions of Kubernetes. Accessing and processing new kinds of resources may have consequences. And there are interaction between functionalities and options to consider, such as how to deal with caching which the exporter already does.
Basically, can we really do this right. Adding a new feature is nice but it has to be reliable and do what users expect with the least amount of caveats. It's still unclear where we can stand not having the Kubernetes API providing Secret reference. This scanning approach comes with so many downsides it questions the relevance of having the exporter perform application mapping.

Having said all that, my initial opinion of doing the mapping on the Prometheus/Alertmanager side is strengthened.
You probably have a source of truth in the form of a CI or an operator where the mapping between Secret and application is known. Having this component also affect the Prometheus/Alertmanager configuration would probably do what you're looking for. Perhaps with rewriting rules, or per-application alerting rules. This could even be compatible with the Prometheus operator and maybe manage with PrometheusRule and AlermanagerConfig resources if these are tools you rely on.

This issue is being discussed at @enix already, keeping you in touch!
Let us know if

**Comment by AnuragEkkati** [comment]
Hello @npdgm 

Interested to know any outcome of the previous comment ? You said it is being discussed at @enix

**Comment by npdgm** [comment]
Sorry for the silence on this issue. Truth be told, the codebase was hard to maintain and prone to regressions, and we lacked the test coverage needed to work on the project with confidence. This contributed to the lack of development. A new version was released last week with a complete architecture overhaul, new features, and solid test suites. Which now puts us in a better position to iterate on feature requests.

Regarding your original request for application mapping information, surfacing object labels is now possible through the Helm chart values `exposeSecretLabels` and `exposeConfigMapLabels`. This could help identify an Helm release for instance, or a Flux Kustomization.
